diesel_codegen_syntex/src/insertable.rs

```diff
@@ -1,84 +1,42 @@
-use syntax::ast::{
-    self,
-    Item,
-    MetaItem,
-    MetaItemKind,
-};
+use syntax::ast::{self, MetaItem};
 use syntax::codemap::Span;
 use syntax::ext::base::{Annotatable, ExtCtxt};
 use syntax::ptr::P;
-use syntax::parse::token::{InternedString, str_to_ident};
 
-use attr::Attr;
+use model::Model;
 use util::{lifetime_list_tokens, struct_ty};
 
-pub fn expand_insert(
+pub fn expand_derive_insertable(
     cx: &mut ExtCtxt,
     span: Span,
-    meta_item: &MetaItem,
+    _meta_item: &MetaItem,
     annotatable: &Annotatable,
     push: &mut FnMut(Annotatable)
 ) {
-    if let Annotatable::Item(ref item) = *annotatable {
-        let tables = insertable_tables(cx, meta_item);
-        for body in tables.into_iter().filter_map(|t| insertable_impl(cx, span, t, item)) {
-            push(Annotatable::Item(body));
-        }
-    } else {
-        cx.span_err(meta_item.span,
-                    "`insertable_into` may only be applied to enums and structs");
-    };
-}
-
-fn insertable_tables(cx: &mut ExtCtxt, meta_item: &MetaItem) -> Vec<InternedString> {
-    match meta_item.node {
-        MetaItemKind::List(_, ref meta_items) => {
-            meta_items.iter().map(|i| table_name(cx, i)).collect()
-        }
-        _ => usage_error(cx, meta_item),
-    }
-}
-
-fn table_name(cx: &mut ExtCtxt, meta_item: &MetaItem) -> InternedString {
-    match meta_item.node {
-        MetaItemKind::Word(ref word) => word.clone(),
-        _ => usage_error(cx, meta_item),
+    if let Some(model) = Model::from_annotable(cx, span, annotatable) {
```

Comment:
Oh, nice. You are basically using the logic you already have in `Model` to do the heavy lifting, right? Actually—is this the exact same code?